judge/views/problem.py

```diff
@@ -213,8 +226,117 @@ def get_context_data(self, **kwargs):
                                           context['description'], 'problem')
         context['meta_description'] = self.object.summary or metadata[0]
         context['og_image'] = self.object.og_image or metadata[1]
+
+        context['can_vote'] = self.can_vote(user, self.object)  # if this problem is votable by this user
+        # the vote this user has already cast on this problem
+        if context['can_vote']:
+            vote = ProblemPointsVote.objects.filter(voter=user.profile, problem=self.object)
+        # whether or not they've already voted
+        context['has_voted'] = context['can_vote'] and vote.exists()
+        if context['has_voted']:
+            context['voted_points'] = vote.first().points  # the previous vote's points
+            context['voted_note'] = vote.first().note
```

Comment:
Each of these (this line and the one above for `context['has_voted']`) is one query to the database. Instead of filtering, you should `ProblemPointsVote.objects.get`, and then catch the `ObjectNotFound` exception.